contracts/tokenbridge/ethereum/L1AtomicTokenBridgeCreator.sol

```diff
@@ -0,0 +1,483 @@
+    // other canonical addresses (dependent on L2 template implementations) can be fetched through `computeExpectedL2***Address` functions
+    address public canonicalL2FactoryAddress;
+    address public canonicalL2ProxyAdminAddress;
+    address public canonicalL2BeaconProxyFactoryAddress;
+
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize() public initializer {
+        __Ownable_init();
+
+        canonicalL2FactoryAddress = _computeAddress(AddressAliasHelper.applyL1ToL2Alias(address(this)), 0);
+        canonicalL2ProxyAdminAddress = Create2.computeAddress(
+            _getL2Salt(OrbitSalts.L2_PROXY_ADMIN), keccak256(type(ProxyAdmin).creationCode), canonicalL2FactoryAddress
+        );
+        canonicalL2BeaconProxyFactoryAddress = Create2.computeAddress(
+            _getL2Salt(OrbitSalts.BEACON_PROXY_FACTORY),
+            keccak256(type(BeaconProxyFactory).creationCode),
+            canonicalL2FactoryAddress
+        );
+    }
+
+    /**
+     * @notice Set addresses of L1 logic contracts and L2 contracts which are deployed on L1.
+     * @dev L2 contracts are deployed to L1 as bytecode placeholders - that bytecode will be part of retryable
+     *      payload used to deploy contracts on L2 side.
+     */
+    function setTemplates(
+        L1GatewayRouter _router,
+        L1ERC20Gateway _standardGateway,
+        L1CustomGateway _customGateway,
+        L1WethGateway _wethGatewayTemplate,
+        address _l2TokenBridgeFactoryTemplate,
+        address _l2RouterTemplate,
+        address _l2StandardGatewayTemplate,
+        address _l2CustomGatewayTemplate,
+        address _l2WethGatewayTemplate,
+        address _l2WethTemplate,
+        address _l1Weth
+    ) external onlyOwner {
+        routerTemplate = _router;
+        standardGatewayTemplate = _standardGateway;
+        customGatewayTemplate = _customGateway;
+        wethGatewayTemplate = _wethGatewayTemplate;
+
+        l2TokenBridgeFactoryTemplate = _l2TokenBridgeFactoryTemplate;
+        l2RouterTemplate = _l2RouterTemplate;
+        l2StandardGatewayTemplate = _l2StandardGatewayTemplate;
+        l2CustomGatewayTemplate = _l2CustomGatewayTemplate;
+        l2WethGatewayTemplate = _l2WethGatewayTemplate;
+        l2WethTemplate = _l2WethTemplate;
+
+        l1Weth = _l1Weth;
+
+        emit OrbitTokenBridgeTemplatesUpdated();
+    }
+
+    /**
+     * @notice Deploy and initialize token bridge, both L1 and L2 sides, as part of a single TX.
+     * @dev This is a single entrypoint of L1 token bridge creator. Function deploys L1 side of token bridge and then uses
+     *      2 retryable tickets  to deploy L2 side. 1st one deploy L2 factory and 2nd calls function that deploys and inits
+     *      all the rest of the contracts. L2 chain is determined by `inbox` parameter.
+     */
+    function createTokenBridge(address inbox, uint256 maxGasForFactory, uint256 maxGasForContracts, uint256 gasPriceBid)
```

Comment:
The L1 factory assumes the L2 factory is deployed at nonce 0 of the L1 factory alias. 

However, since `maxGasForFactory` and  `gasPriceBid` are user controlled, anyone can call `createTokenBridge` with e.g. maxGasForFactory=0 to make the L2 factory deployment fail. The L2 contracts deployment call can still succeed, since it is a call to an empty address. This way, it will burn nonce 0 and the L2 factory can no-longer be deployed at the expected address.

Reply:
A possible fix can be spawning a child contract to send the deployment retryable, so it can never burn the nonce 0 for factory deployment.

Reply:
Given the assumption made, we should also keep track of the nonce per inbox or simply only allow 1 deployment per inbox.

Reply:
Added suggested fix - separate `retryableSender` contract is used to issue retryables which deploy L2 contracts. Also gasLimit for deploying L2 factory gets hard-coded  by owner, in order to avoid burning 0 nonce.